# FRAUD DETECTION in FINANCIAL TRANSACTION

Description: Devlop a machine learnng model to detect fraudulent transaction in a financial transaction in a financial dataset.

Steps:
1.Data Collection: Obtain historical transaction data, including features like transaction amount,    timestamp, etc.
2.Data Preprocessing: Clean the data, handle missing values, and balance the dataset if needed.
3.Feature Engineering: Extract relevant features and engineer new ones that could aid in fraud detection.
4.Exploratory Data Analysis (EDA): Visualize patterns and anomalies in the data using graphs and statistics.
5.Model Selection: Choose classification algorithms like Random Forest, Support Vector Machines, or Neural Networks.
6.Model Training: Train the model on the preprocessed dataset.
7.Model Evaluation: Evaluate the model's performance using metrics like accuracy, precision, recall, and F1-score.

Tech Stack:
Python
Data manipulation libraries 
Machine learning libraries 
Deep learning libraries 



Datasets link: https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud

In [1]:
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn

# warningss
import warnings
warnings.filterwarnings('ignore')

# split data
from sklearn.model_selection import train_test_split

# Linear Regression
from sklearn.linear_model import LinearRegression, LogisticRegression

# Polynomial features
from sklearn.preprocessing import PolynomialFeatures

# calculate accuracy
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score, accuracy_score


In [2]:
df = pd.read_csv("creditcard.csv")
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [3]:
input1 = df.iloc[0, :30].values
input1

array([ 0.00000000e+00, -1.35980713e+00, -7.27811733e-02,  2.53634674e+00,
        1.37815522e+00, -3.38320770e-01,  4.62387778e-01,  2.39598554e-01,
        9.86979013e-02,  3.63786970e-01,  9.07941720e-02, -5.51599533e-01,
       -6.17800856e-01, -9.91389847e-01, -3.11169354e-01,  1.46817697e+00,
       -4.70400525e-01,  2.07971242e-01,  2.57905802e-02,  4.03992960e-01,
        2.51412098e-01, -1.83067779e-02,  2.77837576e-01, -1.10473910e-01,
        6.69280749e-02,  1.28539358e-01, -1.89114844e-01,  1.33558377e-01,
       -2.10530535e-02,  1.49620000e+02])

In [4]:
df.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [6]:
df["Class"].value_counts()


0    284315
1       492
Name: Class, dtype: int64

# Number of legit transactions are: 284315 & illegal transactions are 492
number of legit transactions >> number of illegal transactions



In [7]:
legit_df = df[df["Class"] == 0]
illegal_df = df[df["Class"] == 1]

In [8]:
legit_df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [9]:
illegal_df.head()

Time        V1        V2        V3        V4        V5        V6  \
541    406.0 -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545   
623    472.0 -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823   
4920  4462.0 -2.303350  1.759247 -0.359745  2.330243 -0.821628 -0.075788   
6108  6986.0 -4.397974  1.358367 -2.592844  2.679787 -1.128131 -1.706536   
6329  7519.0  1.234235  3.019740 -4.304597  4.732795  3.624201 -1.357746   

            V7        V8        V9  ...       V21       V22       V23  \
541  -2.537387  1.391657 -2.770089  ...  0.517232 -0.035049 -0.465211   
623   0.325574 -0.067794 -0.270953  ...  0.661696  0.435477  1.375966   
4920  0.562320 -0.399147 -0.238253  ... -0.294166 -0.932391  0.172726   
6108 -3.496197 -0.248778 -0.247768  ...  0.573574  0.176968 -0.436207   
6329  1.713445 -0.496358 -1.282858  ... -0.379068 -0.704181 -0.656805   

           V24       V25       V26       V27       V28  Amount  Class  
541   0.320198  0.044519  0.177840  0.261145 -0.143276    0.00      1  
623  -0.293803  0.279798 -0.145362 -0.252773  0.035764  529.00      1  
4920 -0.087330 -0.156114 -0.542628  0.039566 -0.153029  239.93      1  
6108 -0.053502  0.252405 -0.657488 -0.827136  0.849573   59.00      1  
6329 -1.632653  1.488901  0.566797 -0.010016  0.146793    1.00      1  

[5 rows x 31 columns]

In [10]:
df.groupby("Class").mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

In [11]:
legit_rnd_df = legit_df.sample(1500)
legit_rnd_df.shape

(1500, 31)

In [12]:
new_df = pd.concat([legit_rnd_df, illegal_df])

In [13]:
new_df.groupby("Class").mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      98345.558000  0.069124 -0.019232 -0.033393 -0.039097  0.046062   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.004804  0.073778  0.019696 -0.005172  ...  0.027327 -0.003875   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23      V24       V25       V26       V27       V28  \
Class                                                                        
0      0.007428  0.001709 -0.00446 -0.000105  0.010659  0.002128  0.001730   
1      0.014049 -0.040308 -0.10513  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       91.735820  
1      122.211321  

[2 rows x 30 columns]

# Data splting, training, testing


In [14]:
x = new_df.iloc[:, :30]
x.head()

Time        V1        V2        V3        V4        V5        V6  \
51789    45119.0  0.990473 -0.016380  0.186464  1.135544 -0.153783 -0.301550   
150957   94238.0  2.003866  0.193508 -1.621212  1.399686  0.662103 -0.569879   
243389  151916.0  0.201840  0.814359 -1.039124 -0.351936  1.165946 -0.377859   
180998  124780.0  2.091892 -0.526424 -0.821995 -0.213659 -0.975980 -1.764778   
191625  129346.0  2.065687 -0.146579 -2.143079  0.354943  0.780150 -0.481005   

              V7        V8        V9  ...       V20       V21       V22  \
51789   0.231082 -0.108220 -0.389819  ...  0.169769  0.180257  0.311483   
150957  0.384142 -0.275233  1.397543  ... -0.368948 -0.082141  0.145848   
243389  1.326223 -0.262268 -0.253688  ...  0.196694  0.312094  0.890752   
180998 -0.127472 -0.546304 -0.877305  ... -0.507692 -0.050804  0.470355   
191625  0.484131 -0.302984  0.448681  ... -0.128262 -0.065611 -0.120613   

             V23       V24       V25       V26       V27       V28  Amount  
51789  -0.274221  0.048321  0.672928 -0.293318 -0.002550  0.030613  120.40  
150957 -0.046047 -0.547248  0.423577 -0.518289 -0.053150 -0.083586   14.99  
243389 -0.064932 -0.026465 -0.202777 -0.152459  0.115266  0.139604   94.36  
180998  0.205829  1.004482 -0.146654  0.731170 -0.053927 -0.048809   25.00  
191625 -0.028346  0.013141  0.316781  0.390274 -0.099385 -0.068493   47.54  

[5 rows x 30 columns]

In [15]:
y = new_df.iloc[: , 30]
y.head()

51789     0
150957    0
243389    0
180998    0
191625    0
Name: Class, dtype: int64

In [16]:
x = x.values
y = y.values

In [17]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.1, random_state = 5)

In [18]:
reg = LogisticRegression()

In [19]:
reg.fit(x_train, y_train)

LogisticRegression()

In [20]:
y_pred = reg.predict(x_test)

In [21]:
reg.predict([input1])

array([0], dtype=int64)

In [22]:
accuracy_score(y_pred, y_test)

0.97

# So, accuracy of predicting score is 97.5%